### ISSUE TITLE
@EqualsAndHashCode should allow inclusion of outer class.

### ISSUE THREAD

**Issue by raphw** [rzwitserloot/lombok#1277]
The `@EqualsAndHashCode` annotation should allow the inclusion of outer classes, i.e. for a class:

```java
class Foo {
  @EqualsAndHashCode(includeOuter = true)
  class Bar { }
}
```
Lombok should allow creating `hashCode` and `equals` methods like the following:

```java
class Foo {
  @EqualsAndHashCode(includeOuter = true)
  class Bar { 
    private Foo get$This() {
      return Foo.this;
    }
    @Override int hashCode() { return get$This().hashCode(); }
    @Override boolean equals(Object o) { return o instanceof Bar && get$This().equals(((Bar) o).getThis()); }
  }
}
```

Currently, the only option with Lombok is to explicitly nest the classes what is rather inconvenient.

**Comment by rzwitserloot** [comment]
non-static inner classes are a mistake pretty much at any point, but, hey, if you're going to make this mistake, it's kinda annoying that lombok does the wrong thing.

I don't think includeOuter=true is proper here. Lombok should just include Outer.this as part of the story. 2 objects are not equal if all their fields are entirely equal, but the object is an instance of a (non-static) inner class and the 'owners' aren't the same. Simple as that.

Except, not.

I bet lots of programmers don't actually _know_ that non-static inners have a hidden field containing their outer, which leads to the __disaster__ that fixing this bug will, pragmatically speaking, cause tons of StackOverflowErrors.

Because I expect this to be somewhat commonplace:

    @EqualsAndHashCode
    public class Outer {
        private List<Inner> inners;

        public class Inner {}
    }

Looks simple. Looks innocuous. But this equals method will die with a stack overflow.

It gets worse: Lombok isn't new; if we fix this bug, the above did work, with lombok, before, and now all of a sudden it goes down in flames.

One way out is to warn (and NOT include outer) if we detect that it's a non-static inner, forcing the coder to be explicit about it: add `includeOuter=true` to include the outer, and add `includeOuter=false` to eliminate the warning.

I think that's the best way to solve this bug.
